### ISSUE TITLE
Example cross entropy loss, wrong scaling

### ISSUE THREAD

**Issue by nowozin** [google/jax#1023]
The current examples of using a cross entropy loss, e.g. in https://github.com/google/jax/blob/d90993f798ff00404e2d5869f78d3642d99b6a33/examples/mnist_classifier.py#L43 is

```python
def loss(params, batch):
  inputs, targets = batch
  preds = predict(params, inputs)
  return -np.mean(preds * targets)
```

Due to `np.mean` this code introduces a scaling of (1/class_count) to the loss computation.  I believe the common/correct scaling should be

```python
def loss(params, batch):
  inputs, targets = batch
  preds = predict(params, inputs)
  return -np.mean(np.sum(preds * targets, axis=1))
```

Also, this may be a silly question, but does the multiplication with 0 have the potential to introduce numerical issues upon gradient computation?

**Comment by nowozin** [comment]
Here is an alternative that avoids multiplication with zeros:

```python
def cross_entropy(logprobs, targets):
  target_class = np.argmax(targets, axis=1)
  nll = np.take_along_axis(logprobs, np.expand_dims(target_class, axis=1), axis=1)
  ce = -np.mean(nll)
  return ce

def loss(params, batch):
  inputs, targets = batch
  preds = predict(params, inputs)
  return cross_entropy(preds, targets)
```

**Comment by jonasrauber** [comment]
Just noticed this as well. This should indeed be fixed.

**Comment by jekbradbury** [comment]
Want to send a PR?

**Comment by jonasrauber** [comment]
@jekbradbury done!

P.S. it also improves the training a lot (see PR)
